shared/js/toc.jsx

```diff
@@ -0,0 +1,772 @@
+      sharedMixins.DocumentTitleMixin
+    ],
+
+    propTypes: {
+      // We pass conversationStore here rather than use the mixin, to allow
+      // easy configurability for the ui-showcase.
+      activeRoomStore: React.PropTypes.instanceOf(loop.store.ActiveRoomStore).isRequired,
+      audio: React.PropTypes.object.isRequired,
+      dispatcher: React.PropTypes.instanceOf(loop.Dispatcher).isRequired,
+      introSeen: React.PropTypes.bool,
+      isFirefox: React.PropTypes.bool.isRequired,
+      leaveRoom: React.PropTypes.func.isRequired,
+      // The poster URLs are for UI-showcase testing and development
+      localPosterUrl: React.PropTypes.string,
+      remotePosterUrl: React.PropTypes.string,
+      roomState: React.PropTypes.string,
+      video: React.PropTypes.object.isRequired
+    },
+
+    getInitialState: function() {
+      // Uncomment this line to see the slideshow every time while developing:
+      // localStorage.removeItem("introSeen");
+
+      // Used by the UI showcase to override localStorage value to hide or show FTU slideshow.
+      // localStorage sourced data is always string or null
+      var introSeen = false;
+      if (this.props.introSeen !== undefined) {
+        if (this.props.introSeen) {
+          introSeen = true;
+        }
+      } else {
+        if (localStorage.getItem("introSeen") !== null) {
+          introSeen = true;
+        }
+      }
+      var storeState = this.props.activeRoomStore.getStoreState();
+      return _.extend({}, storeState, {
+        // Used by the UI showcase.
+        roomState: this.props.roomState || storeState.roomState,
+        introSeen: introSeen
+      });
+    },
+
+    componentWillMount: function() {
+      this.props.activeRoomStore.on("change", function() {
+        this.setState(this.props.activeRoomStore.getStoreState());
+      }, this);
+    },
+
+    componentWillUnmount: function() {
+      this.props.activeRoomStore.off("change", null, this);
+    },
+
+    /**
+     * Watches for when we transition to MEDIA_WAIT room state, so we can request
+     * user media access.
+     *
+     * @param  {Object} nextProps (Unused)
+     * @param  {Object} nextState Next state object.
+     */
+    componentWillUpdate: function(nextProps, nextState) {
+      if (this.state.roomState !== ROOM_STATES.READY &&
+          nextState.roomState === ROOM_STATES.READY) {
+        var roomName = nextState.roomName;
+        if (!roomName && nextState.roomContextUrls) {
+          roomName = nextState.roomContextUrls[0].description ||
+              nextState.roomContextUrls[0].location;
+        }
+        if (!roomName) {
+          this.setTitle(mozL10n.get("clientShortname2"));
+        } else {
+          this.setTitle(mozL10n.get("standalone_title_with_room_name", {
+            roomName: roomName,
+            clientShortname: mozL10n.get("clientShortname2")
+          }));
+        }
+      }
+
+      console.log("nextState.roomState", nextState.roomState);
```

Comment:
There's a few of these throughout the patch which we should mark up for removal or just drop.